### DISCUSSION TITLE
Import of external mesh file (.msh etc.)

### DISCUSSION THREAD

**Discussion by SilberVogel** [BriefFiniteElementNet/BriefFiniteElement.Net#140] [Q&A]
Good day for all, is there possibility to import and use external mesh file for this code?

**Comment by epsi1on** [comment]
Hi.
It could added if the mesh file is a well known format. 
Any document or standard etc.?

**Reply by SilberVogel** [reply]
Thx for answer. Standard formats for ANSYS export (.msh, .prj), maybe .stl or text formats...

And how to do it?

**Comment by epsi1on** [comment]
It would not be hard.
To start we must know the syntax of .msh file etc so we need some document or reference etc.

**Reply by SilberVogel** [reply]
Some of shell mesh.


If needed different format, I will send it

**Reply by epsi1on** [reply]
Thanks for the file. I think we also need the .msh file (assuming we want to add support for `.msh` file input)

**Reply by SilberVogel** [reply]
.msh is not necessary at all. I just want understand how to add external mesh in already supporting formats for this soft. Can you list this formats and show example how?

**Comment by epsi1on** [comment]
It is simple. Just create a class named "MshFileImport", put a method "Import()" and write the import code inside...
Then there is need for minor documentation and done...

**Reply by SilberVogel** [reply]
Thanks for explain! If you have any example of project in Viual Studio formats (like .sln) with "Import", please, send it for more understanding.

**Reply by SilberVogel** [reply]
If your code need specific mesh (shell or solid) in specific formats (.db; .inp; .dat; .inc; .uns; .k; .key; .bdf; .nas; .msh; .cas), for make example of "Import" code, tell me, I`ll make it.

**Reply by epsi1on** [reply]
Any of these would be good I think. Can you please suggest one? simple and widely used one. :)

**Reply by SilberVogel** [reply]
Many thanks, .msh format


And last question, does it support shell meshes or only solids?

**Reply by epsi1on** [reply]
good. Do you know any white sheet or definition or standard something that do describe this `.msh` format?

**Comment by epsi1on** [comment]
These are present elements:
1D: BarElement (2node)
2D: Triangle (3node) and Quadrilateral (4node)
3D: Tetrahedral (4node) and Hexahedral (brick element, 8 node)
I think both shell and solids. Not sure exactly what you mean.

**Reply by SilberVogel** [reply]
.msh proprietary file format. Some of I can find about it Im place in archive. In .pdf see Appendix B.

 If any problems with this format, please, tell me.


And what is the next step to start calculations with my own mesh? Please, show me on the example of a Visual Studio project.

**Comment by epsi1on** [comment]
Thanks for the link. its a large file i need to see on PC which will do today or tommorrow. I just found a link from google, is it right format?

http://www.manpagez.com/info/gmsh/gmsh-2.2.6/gmsh_63.php

Next step is to make a class and implement the code. I just created the empty class:
https://github.com/BriefFiniteElementNet/BriefFiniteElement.Net/blob/master/Sources/BriefFiniteElementNet/Utils/MeshImport.cs

**Reply by SilberVogel** [reply]
Yes, think its right, .msh has ASCII encoding.

I saw your changes "Sources/BriefFiniteElementNet/Utils/MeshImport.cs". Wait for your code, and try to do it too.

**Reply by epsi1on** [reply]
ok, so I made an issue for this item (#141)
just need an example .msh file to go. @SilberVogel can you please put an example `.msh` file? thanks

**Reply by SilberVogel** [reply]
Not too big?

**Reply by epsi1on** [reply]
right size but seems wrong format. (have some non ascii characters in the content)
![msh-content](https://user-images.githubusercontent.com/7839141/177577376-ae73990f-7067-4f1a-8fba-6f07c4470b32.png)
Are you sure it is exactly the version that you mean?
Thanks

**Comment by SilberVogel** [comment]
I think its Binary option of export problem, make new files in .msh, .dat and .inp  Maybe if you can’t open it, then you should try later with a different file or format... 
[Mech.zip](https://github.com/BriefFiniteElementNet/BriefFiniteElement.Net/files/9056011/Mech.zip)

I decided to create a code with my simple (Cartesian mesh with 5 elements) model (a console with one fixed wall and a load applied to the other edge) according to the coordinates of the nodes. 


But I found that I can't add more than two nodes to the model for the BarElement class. I understand that a large (100000+) number of nodes is possible only for the TetrahedronElement class? Tell me please, how can I implement the code for my model (I can make Tetrahedron mesh for it). My code:
[BfeTestApplication.zip](https://github.com/BriefFiniteElementNet/BriefFiniteElement.Net/files/9055831/BfeTestApplication.zip)
Many thanks!

**Reply by epsi1on** [reply]
each finite element is related to one bfe element. for above image there should be 5 distinct elements, but you only define one bar element. 
There is some documentation for available elements:
https://bfenet.readthedocs.io/en/latest/elements/finiteElements/index.html
on the left panel of this you can see documentation for different elements

[bar](https://bfenet.readthedocs.io/en/latest/elements/finiteElements/Bar/index.html)
[triangle](https://bfenet.readthedocs.io/en/latest/elements/finiteElements/Triangle/index.html)
[tetrahedron](https://bfenet.readthedocs.io/en/latest/elements/finiteElements/Tetrahedron/index.html)

**Reply by SilberVogel** [reply]
Im trying Tetra mesh, but VS show me errors when Im add TetrahedronElement class... Where Im wrong?

![Tetra_4arguments](https://user-images.githubusercontent.com/98027918/177959174-2b28e716-b2fe-41ec-a7a5-7ac37a092470.png)

Also send .sln project

[Tetra.zip](https://github.com/BriefFiniteElementNet/BriefFiniteElement.Net/files/9070667/Tetra.zip)

**Reply by epsi1on** [reply]
you can do like this:
```
var e1=new TetraHedronElement();
e1.Nodes[0]=n1;e1.Nodes[1]=n25;e1.Nodes[2]=n12;e1.Nodes[3]=n7;
```

**Comment by SilberVogel** [comment]
Question is about code. What I'm missing, especially in material properties (Young and Shear modulus). And do I need calc moments of inertia and set it? It's the same model/mesh:
![Tetra](https://user-images.githubusercontent.com/98027918/178155303-eceb7121-6f3d-4385-bb9b-950058cf9242.png)
[Tetra.zip](https://github.com/BriefFiniteElementNet/BriefFiniteElement.Net/files/9079542/Tetra.zip)

**Reply by epsi1on** [reply]
Hi
Does the compiler show you any error? pretty much like red lines in your visual studio screenshot?
I think the way you define material is not right. can you please send and screenshot like your last screenshot?
thanks

**Comment by SilberVogel** [comment]
Sure)
![Tetra_](https://user-images.githubusercontent.com/98027918/178319938-a4a907a5-8891-4532-8f32-5266557e1a95.png)

**Reply by epsi1on** [reply]
thanks as you can see the error message, it says the `TetrahedronElement` do not have properties of `E` and `G`. This is how you should define the material:

```
var e = 210e9;//210 gpa
var nu = 0.3;

var steelMat = UniformIsotropicMaterial.CreateFromYoungPoisson(e, nu);

e1.Material = e2.Material = steelMat;
```

**Comment by SilberVogel** [comment]
How to chek Internal Forces in each element for Tetrahedron class? Didnt find it in TetrahedronElement.cs

**Comment by epsi1on** [comment]
Yes you are right. it is not implemented yet. Added and issue for it #142 .
Hopefully will be added this week...

**Comment by SilberVogel** [comment]
If I can help with it, tell me

**Reply by epsi1on** [reply]
yes sure, are you familiar with ABAQUS software? do you have access to the software to analyze some models?
I'm asking because, after implementation there is need for the verification of the result. So need to compaire result from this library with result from another (like ansys, abaqus, etc...).
Fortunately prior contributors did it before, you can see result here:
https://github.com/BriefFiniteElementNet/BriefFiniteElement.Net/tree/master/Validation/Validation/Case_03
but i was not able to see the internal force in the results. Unfortunately I moved large amount of files (
4d7a9de) thus did some mess, so lost track of who made the above model. 
All in all we'll need internal force for above model's elements...

**Reply by SilberVogel** [reply]
Only with ANSYS, but in mechanical Im rookie. 

Im check Case_03 and doses not found materials, boundary conditions and forces in .cs file.

**Reply by epsi1on** [reply]
> Only with ANSYS, but in mechanical Im rookie.
> 
> Im check Case_03 and doses not found materials, boundary conditions and forces in .cs file.

material is `E=210e9 Pas` and Poisson's module is 0.25

https://github.com/BriefFiniteElementNet/BriefFiniteElement.Net/blob/9454246beab16ccc95faf1a16ce379ef713cdd43/Validation/Validation/Case_03/Validator.cs#L30

For boundary condition take a look at images, it is simply locked to ground from one side...
the .inp file in the directory is plain text and contains model definition

**Reply by epsi1on** [reply]
there is also another model done by ansys available from here:
https://github.com/BriefFiniteElementNet/BriefFiniteElement.Net/tree/master/Validation/Validation/Data/Tet4-78knv
We can use it for validation but if validation fails, then we should debug with a smaller model (a model like Case_3 which have few elements)

**Reply by SilberVogel** [reply]
Results for Case_3 in ANSYS
[Case 3 Stress for Brief.zip](https://github.com/BriefFiniteElementNet/BriefFiniteElement.Net/files/9120793/Case.3.Stress.for.Brief.zip)

**Reply by epsi1on** [reply]
Wonderful.
Is it identical model with exact node location and element-node connection?
If yes then it should report exactly same nodal displacement with current case_3 but i did not see the nodal displacements in order to check it.
By the way i'll use it for validation.

Thanks again

**Reply by epsi1on** [reply]
If you mind, can you please create a pull request and put these output files into [Case_3 folder](https://github.com/BriefFiniteElementNet/BriefFiniteElement.Net/tree/master/Validation/Validation/Case_03)?
This way your username will be listed in project history, and make it simpler to track the creator of file in future. 
If you have any issue with creating the pull request, then please let me know, I'll add them myself ...

Thanks.

**Reply by SilberVogel** [reply]
Sorry, Im newbie at GH... Better you put them.

I`ll chek displacements.

**Reply by SilberVogel** [reply]
> Wonderful. Is it identical model with exact node location and element-node connection? If yes then it should report exactly same nodal displacement with current case_3 but i did not see the nodal displacements in order to check it. By the way i'll use it for validation.
> 
> Thanks again

Yes, for this case were used Job-10.inp file from .../BriefFiniteElement.Net/Validation/Case_03/

[Case 3 Deformation for Brief.zip](https://github.com/BriefFiniteElementNet/BriefFiniteElement.Net/files/9131647/Case.3.Deformation.for.Brief.zip)

**Reply by epsi1on** [reply]
As you know the strain and stress in constant through a tetrahedron element, But as long as i see, in your output, the stress is not constant through tethrahedrons. Just thinking about the reason... Do you have any idea?

Can we please continue discussion about tetra hedron internal force here: #142 (for future reference)
Thanks

**Comment by epsi1on** [comment]
inactivity
